https://github.com/LibreOffice/core/commit/0e7d4d99beda181f4e22725e4b8295d25489696f
https://bugs.gentoo.org/821265

From 0e7d4d99beda181f4e22725e4b8295d25489696f Mon Sep 17 00:00:00 2001
From: Stephan Bergmann <sbergman@redhat.com>
Date: Sun, 22 Aug 2021 11:26:38 +0200
Subject: [PATCH] gengal needs Library_localedata_en now

Whatever caused that requirement, but a from-scratch build of e.g.
Gallery_backgrounds failed for me (with a follow-on SIGABRT after
"framework/source/services/desktop.cxx:176: Desktop::~Desktop(): Who forgot to
dispose this service?") because it failed to load
instdir/program/../program/liblocaledata_en.so at

> #1  0x00007fffda0b9671 in i18npool::LocaleDataImpl::getFunctionSymbol(com::sun::star::lang::Locale const&, char const*) (this=0x159b630, rLocale=..., pFunction=0x7fffda136ed8 "getAllFormats0") at i18npool/source/localedata/localedata.cxx:1460
> #2  0x00007fffda0b6a93 in i18npool::LocaleDataImpl::FormatSection::getFunc(i18npool::LocaleDataImpl&, com::sun::star::lang::Locale const&, char const*) (this=0x7fffffffc840, rLocaleData=..., rL=..., pName=0x7fffda136ed8 "getAllFormats0") at i18npool/source/localedata/localedata.cxx:857
> #3  0x00007fffda0b6b5c in i18npool::LocaleDataImpl::getAllFormats(com::sun::star::lang::Locale const&) (this=0x159b630, rLocale=...) at i18npool/source/localedata/localedata.cxx:865
> #4  0x00007fffda0d4815 in NumberFormatCodeMapper::getFormats(com::sun::star::lang::Locale const&) (this=0x1594fa0, rLocale=...) at i18npool/source/numberformatcode/numberformatcode.cxx:150
> #5  0x00007fffda0d4155 in NumberFormatCodeMapper::getAllFormatCode(short, com::sun::star::lang::Locale const&) (this=0x1594fa0, formatUsage=8, rLocale=...) at i18npool/source/numberformatcode/numberformatcode.cxx:92
> #6  0x00007ffff572a824 in LocaleDataWrapper::loadCurrencyFormats() (this=0x159b3f0) at unotools/source/i18n/localedatawrapper.cxx:544
> #7  0x00007ffff572777e in LocaleDataWrapper::loadData() (this=0x159b3f0) at unotools/source/i18n/localedatawrapper.cxx:131
> #8  0x00007ffff57271b4 in LocaleDataWrapper::LocaleDataWrapper(LanguageTag const&, std::__debug::vector<rtl::OUString, std::allocator<rtl::OUString> > const&) (this=0x159b3f0, rLanguageTag=..., rOverrideDateAcceptancePatterns=std::__debug::vector of length 0, capacity 0) at unotools/source/i18n/localedatawrapper.cxx:80
> #9  0x00007ffff5766fd6 in SvtSysLocale_Impl::SvtSysLocale_Impl() (this=0x158a4e0) at unotools/source/misc/syslocale.cxx:67
> #10 0x00007ffff5768ad7 in std::construct_at<SvtSysLocale_Impl>(SvtSysLocale_Impl*) (__location=0x158a4e0) at ~/gcc/trunk/inst/include/c++/12.0.0/bits/stl_construct.h:97
> #11 0x00007ffff5768b1c in std::allocator_traits<std::allocator<SvtSysLocale_Impl> >::construct<SvtSysLocale_Impl>(std::allocator<SvtSysLocale_Impl>&, SvtSysLocale_Impl*) (__a=..., __p=0x158a4e0) at ~/gcc/trunk/inst/include/c++/12.0.0/bits/alloc_traits.h:514
> #12 0x00007ffff57688c2 in std::_Sp_counted_ptr_inplace<SvtSysLocale_Impl, std::allocator<SvtSysLocale_Impl>, (__gnu_cxx::_Lock_policy)2>::_Sp_counted_ptr_inplace<>(std::allocator<SvtSysLocale_Impl>) (this=0x158a4d0, __a=...) at ~/gcc/trunk/inst/include/c++/12.0.0/bits/shared_ptr_base.h:519
> #13 0x00007ffff576860a in std::__shared_count<(__gnu_cxx::_Lock_policy)2>::__shared_count<SvtSysLocale_Impl, std::allocator<SvtSysLocale_Impl>>(SvtSysLocale_Impl*&, std::_Sp_alloc_shared_tag<std::allocator<SvtSysLocale_Impl> >) (this=0x7fffffffd0a8, __p=@0x7fffffffd0a0: 0x0, __a=...) at ~/gcc/trunk/inst/include/c++/12.0.0/bits/shared_ptr_base.h:650
> #14 0x00007ffff57684c6 in std::__shared_ptr<SvtSysLocale_Impl, (__gnu_cxx::_Lock_policy)2>::__shared_ptr<std::allocator<SvtSysLocale_Impl>>(std::_Sp_alloc_shared_tag<std::allocator<SvtSysLocale_Impl> >) (this=0x7fffffffd0a0, __tag=...) at ~/gcc/trunk/inst/include/c++/12.0.0/bits/shared_ptr_base.h:1342
> #15 0x00007ffff576838d in std::shared_ptr<SvtSysLocale_Impl>::shared_ptr<std::allocator<SvtSysLocale_Impl>>(std::_Sp_alloc_shared_tag<std::allocator<SvtSysLocale_Impl> >) (this=0x7fffffffd0a0, __tag=...) at ~/gcc/trunk/inst/include/c++/12.0.0/bits/shared_ptr.h:409
> #16 0x00007ffff5768164 in std::allocate_shared<SvtSysLocale_Impl, std::allocator<SvtSysLocale_Impl>>(std::allocator<SvtSysLocale_Impl> const&) (__a=...) at ~/gcc/trunk/inst/include/c++/12.0.0/bits/shared_ptr.h:863
> #17 0x00007ffff5767dfc in std::make_shared<SvtSysLocale_Impl>() () at ~/gcc/trunk/inst/include/c++/12.0.0/bits/shared_ptr.h:879
> #18 0x00007ffff5767583 in SvtSysLocale::SvtSysLocale() (this=0x7fffffffd120) at unotools/source/misc/syslocale.cxx:122
> #19 0x00007ffff015148f in FwkResId(TranslateId) (aId=...) at framework/source/fwe/classes/fwkresid.cxx:22
> #20 0x00007ffff025f333 in framework::Desktop::constructorInit() (this=0x1588850) at framework/source/services/desktop.cxx:106
> #21 0x00007ffff0265c1c in (anonymous namespace)::createDesktop(com::sun::star::uno::Reference<com::sun::star::uno::XComponentContext> const&) (context=...) at framework/source/services/desktop.cxx:1764
> #22 0x00007ffff0265ce5 in framework::getDesktop(com::sun::star::uno::Reference<com::sun::star::uno::XComponentContext> const&) (context=...) at framework/source/services/desktop.cxx:1773
> #23 0x00007ffff0265d9e in com_sun_star_comp_framework_Desktop_get_implementation(com::sun::star::uno::XComponentContext*, com::sun::star::uno::Sequence<com::sun::star::uno::Any> const&) (context=0x522200) at framework/source/services/desktop.cxx:1782
> #24 0x00007ffff5a5d864 in std::__invoke_impl<com::sun::star::uno::XInterface*, com::sun::star::uno::XInterface* (*&)(com::sun::star::uno::XComponentContext*, com::sun::star::uno::Sequence<com::sun::star::uno::Any> const&), com::sun::star::uno::XComponentContext*, com::sun::star::uno::Sequence<com::sun::star::uno::Any> const&>(std::__invoke_other, com::sun::star::uno::XInterface* (*&)(com::sun::star::uno::XComponentContext*, com::sun::star::uno::Sequence<com::sun::star::uno::Any> const&), com::sun::star::uno::XComponentContext*&&, com::sun::star::uno::Sequence<com::sun::star::uno::Any> const&) (__f=@0x54bb58: 0x7ffff0265d5f <com_sun_star_comp_framework_Desktop_get_implementation(com::sun::star::uno::XComponentContext*, com::sun::star::uno::Sequence<com::sun::star::uno::Any> const&)>) at ~/gcc/trunk/inst/include/c++/12.0.0/bits/invoke.h:61
> #25 0x00007ffff5a551f4 in std::__invoke_r<com::sun::star::uno::XInterface*, com::sun::star::uno::XInterface* (*&)(com::sun::star::uno::XComponentContext*, com::sun::star::uno::Sequence<com::sun::star::uno::Any> const&), com::sun::star::uno::XComponentContext*, com::sun::star::uno::Sequence<com::sun::star::uno::Any> const&>(com::sun::star::uno::XInterface* (*&)(com::sun::star::uno::XComponentContext*, com::sun::star::uno::Sequence<com::sun::star::uno::Any> const&), com::sun::star::uno::XComponentContext*&&, com::sun::star::uno::Sequence<com::sun::star::uno::Any> const&) (__fn=@0x54bb58: 0x7ffff0265d5f <com_sun_star_comp_framework_Desktop_get_implementation(com::sun::star::uno::XComponentContext*, com::sun::star::uno::Sequence<com::sun::star::uno::Any> const&)>) at ~/gcc/trunk/inst/include/c++/12.0.0/bits/invoke.h:114
> #26 0x00007ffff5a4d254 in std::_Function_handler<com::sun::star::uno::XInterface* (com::sun::star::uno::XComponentContext*, com::sun::star::uno::Sequence<com::sun::star::uno::Any> const&), com::sun::star::uno::XInterface* (*)(com::sun::star::uno::XComponentContext*, com::sun::star::uno::Sequence<com::sun::star::uno::Any> const&)>::_M_invoke(std::_Any_data const&, com::sun::star::uno::XComponentContext*&&, com::sun::star::uno::Sequence<com::sun::star::uno::Any> const&) (__functor=..., __args#0=@0x7fffffffd4f0: 0x522200, __args#1=...) at ~/gcc/trunk/inst/include/c++/12.0.0/bits/std_function.h:291
> #27 0x00007ffff5a3f913 in std::function<com::sun::star::uno::XInterface* (com::sun::star::uno::XComponentContext*, com::sun::star::uno::Sequence<com::sun::star::uno::Any> const&)>::operator()(com::sun::star::uno::XComponentContext*, com::sun::star::uno::Sequence<com::sun::star::uno::Any> const&) const (this=0x54bb58, __args#0=0x522200, __args#1=...) at ~/gcc/trunk/inst/include/c++/12.0.0/bits/std_function.h:568
> #28 0x00007ffff5a2a396 in cppuhelper::ServiceManager::Data::Implementation::doCreateInstance(com::sun::star::uno::Reference<com::sun::star::uno::XComponentContext> const&) (this=0x54baa0, context=...) at cppuhelper/source/servicemanager.cxx:704
> #29 0x00007ffff5a2a12a in cppuhelper::ServiceManager::Data::Implementation::createInstance(com::sun::star::uno::Reference<com::sun::star::uno::XComponentContext> const&, bool) (this=0x54baa0, context=..., singletonRequest=false) at cppuhelper/source/servicemanager.cxx:672
> #30 0x00007ffff5a2d02d in cppuhelper::ServiceManager::createInstanceWithContext(rtl::OUString const&, com::sun::star::uno::Reference<com::sun::star::uno::XComponentContext> const&) (this=0x523450, aServiceSpecifier=..., Context=...) at cppuhelper/source/servicemanager.cxx:1003
> #31 0x00007ffff6cd9922 in com::sun::star::frame::Desktop::create(com::sun::star::uno::Reference<com::sun::star::uno::XComponentContext> const&) (the_context=...) at workdir/UnoApiHeadersTarget/offapi/normal/com/sun/star/frame/Desktop.hpp:38
> #32 0x00007ffff6cd8c07 in SfxApplication::Initialize_Impl() (this=0x1588150) at sfx2/source/appl/appinit.cxx:182
> #33 0x00007ffff6cb776a in SfxApplication::GetOrCreate() () at sfx2/source/appl/app.cxx:122
> #34 0x0000000000405ecf in (anonymous namespace)::GalApp::Main() (this=0x4204e0 <vclmain::createApplication()::aGalApp>) at svx/source/gengal/gengal.cxx:254
> #35 0x00007ffff2dd6777 in ImplSVMain() () at vcl/source/app/svmain.cxx:199
> #36 0x00007ffff2dd6898 in SVMain() () at vcl/source/app/svmain.cxx:231
> #37 0x0000000000410f1f in sal_main() () at vcl/source/salmain/salmain.cxx:34
> #38 0x0000000000410efa in main(int, char**) (argc=10, argv=0x7fffffffded8) at vcl/source/salmain/salmain.cxx:29

This fix should remove the need for the cross-build--only fix
0e142fa01972ad28182ca54b855d0ecae0d5c83b "cross-toolset: gengal now needs
localedata_en", which it reverts.

Change-Id: If71b7249c19619a83911c7624e3d99077decba31
Reviewed-on: https://gerrit.libreoffice.org/c/core/+/120837
Tested-by: Jenkins
Reviewed-by: Stephan Bergmann <sbergman@redhat.com>
---
 RepositoryExternal.mk                      | 1 +
 solenv/gbuild/extensions/pre_BuildTools.mk | 1 -
 2 files changed, 1 insertion(+), 1 deletion(-)

diff --git a/RepositoryExternal.mk b/RepositoryExternal.mk
index 6e732c94793b2..15cfbfde3d4ac 100644
--- a/RepositoryExternal.mk
+++ b/RepositoryExternal.mk
@@ -3875,6 +3875,7 @@ endef
 define gb_Executable__register_gengal
 $(call gb_Executable_add_runtime_dependencies,gengal,\
 	$(call gb_Library_get_target_for_build,$(CPPU_ENV_FOR_BUILD)_uno) \
+	$(call gb_Library_get_target_for_build,localedata_en) \
 	$(if $(filter MACOSX,$(OS_FOR_BUILD)),$(call gb_Library_get_target_for_build,vclplug_osx)) \
 	$(if $(filter WNT,$(OS_FOR_BUILD)),$(call gb_Library_get_target_for_build,vclplug_win)) \
 	$(if $(filter host,$(gb_Side)),$(call gb_Package_get_target,postprocess_images)) \
diff --git a/solenv/gbuild/extensions/pre_BuildTools.mk b/solenv/gbuild/extensions/pre_BuildTools.mk
index e08906068d246..84e2ea533673f 100644
--- a/solenv/gbuild/extensions/pre_BuildTools.mk
+++ b/solenv/gbuild/extensions/pre_BuildTools.mk
@@ -47,7 +47,6 @@ gb_BUILD_TOOLS = \
     $(foreach executable,$(gb_BUILD_TOOLS_executables),$(call gb_Executable_get_runtime_dependencies,$(executable))) \
     $(foreach executable,$(gb_BUILD_TOOLS_executables_extern),$(call gb_ExternalExecutable_get_dependencies,$(executable))) \
 	$(INSTROOT)/$(LIBO_URE_ETC_FOLDER)/$(call gb_Helper_get_rcfile,uno) \
-    $(if $(WITH_GALLERY_BUILD),$(call gb_Library_get_target,localedata_en)) \
 
 endif
 
